crates/accelerate/src/synthesis/multi_controlled/mcmt.rs

```diff
@@ -0,0 +1,181 @@
+    indices.map(|(ctrl1, ctrl2, target)| {
+        (
+            StandardGate::CCXGate.into(),
+            smallvec![],
+            vec![
+                Qubit(ctrl1 as u32),
+                Qubit(ctrl2 as u32),
+                Qubit(target as u32),
+            ],
+            vec![],
+        )
+    })
+}
+
+/// Implement multi-control, multi-target of a single-qubit gate using a V-chain with
+/// (num_ctrl_qubits - 1) auxiliary qubits.
+/// ``controlled_gate`` here must already be the controlled operation, e.g. if we
+/// call MCMT of X, then it must be a CX gate. This is because I currently don't know how to
+/// nicely map the single-qubit gate to it's controlled version.
+///
+/// For example, 4 controls and 2 target qubits for the Hadamard gate, generates
+///
+///     q_0: ──■──────────────────────────────────■──
+///            │                                  │
+///     q_1: ──■──────────────────────────────────■──
+///            │                                  │
+///     q_2: ──┼────■────────────────────────■────┼──
+///            │    │                        │    │
+///     q_3: ──┼────┼────■──────────────■────┼────┼──
+///            │    │    │  ┌───┐       │    │    │
+///     q_4: ──┼────┼────┼──┤ H ├───────┼────┼────┼──
+///            │    │    │  └─┬─┘┌───┐  │    │    │
+///     q_5: ──┼────┼────┼────┼──┤ H ├──┼────┼────┼──
+///          ┌─┴─┐  │    │    │  └─┬─┘  │    │  ┌─┴─┐
+///     q_6: ┤ X ├──■────┼────┼────┼────┼────■──┤ X ├
+///          └───┘┌─┴─┐  │    │    │    │  ┌─┴─┐└───┘
+///     q_7: ─────┤ X ├──■────┼────┼────■──┤ X ├─────
+///               └───┘┌─┴─┐  │    │  ┌─┴─┐└───┘
+///     q_8: ──────────┤ X ├──■────■──┤ X ├──────────
+///                    └───┘          └───┘
+///
+#[pyfunction]
+#[pyo3(signature = (controlled_gate, num_ctrl_qubits, num_target_qubits, control_state=None))]
+pub fn mcmt_v_chain(
+    py: Python,
+    controlled_gate: OperationFromPython,
+    num_ctrl_qubits: usize,
+    num_target_qubits: usize,
+    control_state: Option<usize>,
+) -> PyResult<CircuitData> {
+    let packed_controlled_gate = controlled_gate.operation;
+    let num_qubits = if num_ctrl_qubits > 1 {
+        2 * num_ctrl_qubits - 1 + num_target_qubits
+    } else {
+        1 + num_target_qubits // we can have 1 control and multiple targets
+    };
+
+    let control_state = control_state
+        .or(Some(usize::pow(2, num_ctrl_qubits as u32) - 1))
+        .unwrap();
+
+    // First, we handle bitflips in case of open controls.
+    let flip_control_state = (0..num_ctrl_qubits)
+        .filter(|index| control_state & (1 << index) == 0)
+        .map(|index| {
+            (
+                PackedOperation::from_standard(StandardGate::XGate),
+                smallvec![] as SmallVec<[Param; 3]>,
+                vec![Qubit(index as u32)],
+                vec![] as Vec<Clbit>,
+            )
+        });
+
+    // First, we create the operations that apply the controlled base gate.
```

Comment:
This can no longer be "first" (my internal comment-reader-counter raises an error) 😛

Reply:
You don't count 1 1 2 3? 😛

Reply:
Fixed 🙂